packages/ra-core/src/dataProvider/replyWithCache.ts

```diff
@@ -0,0 +1,61 @@
+import {
+    GetListParams,
+    GetListResult,
+    GetOneParams,
+    GetOneResult,
+    GetManyParams,
+    GetManyResult,
+} from '../types';
+
+export const canReplyWithCache = (type, payload, resourceState) => {
+    const now = new Date();
+    switch (type) {
+        case 'getList':
+            return (
+                resourceState &&
+                resourceState.list &&
+                resourceState.list.validity &&
+                resourceState.list.validity[
+                    JSON.stringify(payload as GetListParams)
+                ] > now
+            );
+        case 'getOne':
+            return (
+                resourceState &&
+                resourceState.validity &&
+                resourceState.validity[(payload as GetOneParams).id] > now
+            );
+        case 'getMany':
+            return (
+                resourceState &&
+                resourceState.validity &&
+                (payload as GetManyParams).ids.every(
+                    id => resourceState.validity[id] > now
+                )
+            );
+        default:
+            return false;
+    }
+};
+
+export const getResultFromCache = (type, payload, resourceState) => {
+    switch (type) {
+        case 'getList': {
+            const data = resourceState.data;
+            const requestSignature = JSON.stringify(payload);
+            const ids = resourceState.list.idsForQuery[requestSignature];
```

Comment:
Strange, are we sure resourceState.list.idsForQuery is always set ?

Reply:
yes: if the query key was set in `validity`, it means it was done in `idsForQuery`, too.